src/charm.py

```diff
@@ -509,73 +525,147 @@ def relabeling_config(self) -> list:
-                        },
-                    }
-                ],
-                "pipeline_stages": [
-                    {
-                        "drop": {
-                            "expression": ".*file is a directory.*",
-                        },
-                    },
-                ],
-            }
+        if self.config["classic_snap"]:
+            for endpoint, topology in self._cos.snap_log_endpoints_with_topology:
+                # endpoint: owner:name
+                try:
+                    with open(f"/snap/{endpoint.owner}/current/meta/snap.yaml") as f:
+                        snap_yaml = yaml.safe_load(f)
+                except FileNotFoundError:
+                    logger.error(
+                        f"snap file for {endpoint.owner} not found. It is likely not installed. Skipping."
+                    )
+                    continue
+                log_dirs = snap_yaml["slots"][endpoint.name]["source"]["read"]
+                for key in snap_yaml["apps"].keys():
+                    snap_app_name = key  # Just use any app.
+                    break
+                log_dirs = self._evaluate_log_paths(
+                    paths=log_dirs, snap=endpoint.owner, app=snap_app_name
+                )
+                for path in log_dirs:
+                    job = self._snap_plug_job(
+                        endpoint.owner,
+                        f"{path}/**",
+                        topology.application,
+                        str(topology.unit),
+                        self._path_label(path),
+                    )
+                    shared_logs_configs.append(job)
 
-            job["relabel_configs"] = [
-                {
-                    "source_labels": ["__path__"],
-                    "target_label": "path",
-                    "replacement": fstab_entry.relative_target,
+        else:
+            endpoint_owners = {
+                endpoint.owner: {
+                    "juju_application": topology.application,
+                    "juju_unit": topology.unit,
                 }
-            ]
-
-            shared_logs_configs.append(job)
+                for endpoint, topology in self._cos.snap_log_endpoints_with_topology
+            }
+            for fstab_entry in agent_fstab.entries:
+                if fstab_entry.owner not in endpoint_owners.keys():
+                    continue
+
+                target_path = (
+                    f"{fstab_entry.target}/**"
+                    if fstab_entry
+                    else "/snap/grafana-agent/current/shared-logs/**"
+                )
+
+                job = self._snap_plug_job(
+                    fstab_entry.owner,
+                    target_path,
+                    endpoint_owners[fstab_entry.owner]["juju_application"],
+                    endpoint_owners[fstab_entry.owner]["juju_unit"],
+                    fstab_entry.relative_target,
+                )
+                shared_logs_configs.append(job)
```

Comment:
Could you please sprinkle some comments throughout the function? I'm having trouble understanding it.